### PR TITLE
daemon/c8d: Implement image history

### PR DESCRIPTION
**Pull request by rumpl**
**- What I did**

Implemented `docker image history`.

Note there is some duplication of code with the usual resolveDescriptor -> images.Config -> content.ReadBlob, I think we can refactor at a later time.

I'm unsure how `Tags` work, there is a difference between an image history when it's built by the classic builder or buildkit:

Currently, with this Dockerfile:
```dockerfile
FROM alpine

CMD ["echo", "hello"]
```

If we build it with the classic builder (``) we get this history:

```json
[
  {
    "Comment": "",
    "Created": 1673531696,
    "CreatedBy": "/bin/sh -c #(nop)  CMD [\"echo\" \"hello\"]",
    "Id": "sha256:42f9b7e41f28aa91037a8c8355de66d650c6fd192426e7d30c33591a5e72e6ea",
    "Size": 0,
    "Tags": [
      "henelo:latest"
    ]
  },
  {
    "Comment": "",
    "Created": 1673283889,
    "CreatedBy": "/bin/sh -c #(nop)  CMD [\"/bin/sh\"]",
    "Id": "sha256:04eeaa5f8c35b8206a15c12425498c37ecd2181e0ef202bea24a2382e4dd240f",
    "Size": 0,
    "Tags": [
      "alpine:latest"
    ]
  },
  {
    "Comment": "",
    "Created": 1673283888,
    "CreatedBy": "/bin/sh -c #(nop) ADD file:3080f19f39259a4b77cc53975de0184c78d4335ceb9ffb77a2838d0539ad6f85 in / ",
    "Id": "<missing>",
    "Size": 7458929,
    "Tags": null
  }
]
```

But if we build it with buidkit we get this:

```json
[
  {
    "Comment": "buildkit.dockerfile.v0",
    "Created": 1673283889,
    "CreatedBy": "CMD [\"echo\" \"hello\"]",
    "Id": "sha256:6f6cafc4e6423f0997d34c88b4e018028c5225c6ccb67f55c2f0efd5f8a808aa",
    "Size": 0,
    "Tags": [
      "henelo:latest"
    ]
  },
  {
    "Comment": "",
    "Created": 1673283889,
    "CreatedBy": "/bin/sh -c #(nop)  CMD [\"/bin/sh\"]",
    "Id": "<missing>",
    "Size": 0,
    "Tags": null
  },
  {
    "Comment": "",
    "Created": 1673283888,
    "CreatedBy": "/bin/sh -c #(nop) ADD file:3080f19f39259a4b77cc53975de0184c78d4335ceb9ffb77a2838d0539ad6f85 in / ",
    "Id": "<missing>",
    "Size": 7458929,
    "Tags": null
  }
]
```

Note how the image that's built with buildkit looses the `alpine:latest` tags in the second history item.

Also, the only way to see these `Tags` is to call the API directly, `docker image history <name> --format '{{ json . }}'` doesn't even show the `Tags` property.

~~Another difference with the exising history API is that the `Tags` will not be `null` since the docs say that `Tags` is required.~~

Upstreams:
* https://github.com/rumpl/moby/pull/69

**- How I did it**

By looping over the `History` of the image.

**- How to verify it**

Pull or build an image and then call `docker image history`.

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**
Screenshot 2023-01-12 at 14 40 37

### CODE REVIEW

**Review by vvoland — APPROVED** [review]

File: daemon/containerd/image_history.go [hunk 1]
```diff
@@ -2,14 +2,104 @@ package containerd
...
+	}
+
+	history := []*imagetype.HistoryResponseItem{}
+
+	sizes := []int64{}
+	s := i.client.SnapshotService(i.snapshotter)
+
+	for i := range diffIDs {
+		diffIDs := diffIDs[0 : i+1]
+		chainID := identity.ChainID(diffIDs).String()
```
Comment (corhere) [inline daemon/containerd/image_history.go:55]: Nit: shadowing `diffIDs` in the loop body is a touch too clever for my taste.
```suggestion
		chainID := identity.ChainID(diffIDs[0:i+1]).String()
```

File: daemon/containerd/image_history.go [hunk 2]
```diff
@@ -2,14 +2,104 @@ package containerd
...
+			sizes = sizes[1:]
+		}
+
+		history = append([]*imagetype.HistoryResponseItem{{
+			ID:        "<missing>",
+			Comment:   h.Comment,
+			CreatedBy: h.CreatedBy,
+			Created:   h.Created.Unix(),
+			Size:      size,
+			Tags:      []string{},
```
Comment (corhere) [inline daemon/containerd/image_history.go:81]:
```suggestion
```

File: daemon/containerd/image_history.go [hunk 3]
```diff
@@ -2,14 +2,104 @@ package containerd
...
+	}
+
+	var image ocispec.Image
+	if err := json.Unmarshal(blob, &image); err != nil {
+		return nil, err
+	}
+
+	history := []*imagetype.HistoryResponseItem{}
+
+	sizes := []int64{}
```
Comment (corhere) [inline daemon/containerd/image_history.go:50]:
```suggestion
	var (
		history []*imagetype.HistoryResponseItem
		sizes   []int64
	)
```
https://github.com/golang/go/wiki/CodeReviewComments#declaring-empty-slices

**Review by corhere — COMMENTED** [review]
LGTM, though I have some nits to pick.

File: daemon/containerd/image_history.go (hunk 2 above)
Comment (rumpl) [reply daemon/containerd/image_history.go:81]: `Tags` is required in the API https://docs.docker.com/engine/api/v1.41/#tag/Image/operation/ImageHistory, the current implementation does indeed return `"Tags": null` but that just feels wrong for two reasons:
* there is no need for a slice to be nil, an empty slice avoids having to check  if it's null
* the field is required

Also, from [[here]()](https://github.com/golang/go/wiki/CodeReviewComments#declaring-empty-slices):
```
Note that there are limited circumstances where a non-nil but zero-length slice is preferred, such as when encoding JSON objects (a nil slice encodes to null, while []string{} encodes to the JSON array []).
```

File: daemon/containerd/image_history.go (hunk 1 above)
Comment (rumpl) [reply daemon/containerd/image_history.go]: Done

File: daemon/containerd/image_history.go (hunk 3 above)
Comment (rumpl) [reply daemon/containerd/image_history.go]: 👍

File: daemon/containerd/image_history.go (hunk 2 above)
Comment (corhere) [reply daemon/containerd/image_history.go:81]: > the current implementation does indeed return `"Tags": null` but that just feels wrong for two reasons:
> 
> * there is no need for a slice to be nil, an empty slice avoids having to check  if it's null

While I agree with you in general, that returning an empty JSON array is more friendly to non-Go clients than `null`, it's rather futile to fix it here and now _because_ of the existing behaviour. Any client which wants to be compatible with any daemon older than v-next needs to be prepared to handle responses containing `"Tags": null`. And it's possible that some client could be using `history.Tags != null` to check if the image has any tags. I would love to make the API more consistent, but for the sake of forwards and backwards compatibility I think this change in behaviour needs to be API version-gated. (Furthermore, promoting the `nil` slice to an empty slice before marshaling to JSON depending on the API version is well within the scope of responsibility for an API handler function, IMO.)

In summary, my recommended course of action is to:
1. change this implementation to return `Tags: nil` to match the existing behaviour, and
2. follow-up with a PR which changes `(*imageRouter).getImagesHistory` to respond with `Tags: []` when API version >= 1.44

**Review by corhere — APPROVED** [review]

File: daemon/containerd/image_history.go (hunk 2 above)
Comment (rumpl) [reply daemon/containerd/image_history.go]: done!

I opened a ticket https://github.com/moby/moby/issues/44872

File: daemon/containerd/image_history.go [hunk 4]
```diff
@@ -2,14 +2,104 @@ package containerd
...
+			Comment:   h.Comment,
+			CreatedBy: h.CreatedBy,
+			Created:   h.Created.Unix(),
+			Size:      size,
+			Tags:      nil,
+		}}, history...)
+	}
+
+	if len(history) != 0 {
+		history[0].ID = desc.Digest.String()
```
Comment (thaJeztah) [inline daemon/containerd/image_history.go:86]: For the graphdriver implementation, we loop through all history entries to get this information. Is it intentional (e.g. not possible) for the snapshotter implementation? https://github.com/moby/moby/blob/65c371ed6e48d0014a4826a6933f9231f688327a/daemon/images/image_history.go#L57-L58

Comment (rumpl) [reply daemon/containerd/image_history.go:86]: The `id` it's taking here is first the id of the image, but then it the id of the parent (https://github.com/moby/moby/blob/65c371ed6e48d0014a4826a6933f9231f688327a/daemon/images/image_history.go#L69) we don't have this relation in containerd unfortunately, or at least I didn't find any way to do it. 

That parent relation doesn't exist if you build the image with buildkit (see the PR description).

We could followup when we have the classic builder working?

File: daemon/containerd/image_history.go [hunk 5]
```diff
@@ -2,14 +2,104 @@ package containerd
...
+		if !h.EmptyLayer {
+			if len(sizes) == 0 {
+				return nil, errors.New("unable to find the size of the layer")
+			}
+			size = sizes[0]
+			sizes = sizes[1:]
+		}
+
+		history = append([]*imagetype.HistoryResponseItem{{
+			ID:        "<missing>",
```
Comment (tianon) [inline daemon/containerd/image_history.go:76]: Another one of @thaJeztah's favorite cases of presentation strings that are part of the API but probably shouldn't be. :joy: :sob: :heart:

File: daemon/containerd/image_history.go [hunk 6]
```diff
@@ -2,14 +2,104 @@ package containerd
...
+
+	blob, err := content.ReadBlob(ctx, cs, conf)
+	if err != nil {
+		return nil, err
+	}
+
+	var image ocispec.Image
+	if err := json.Unmarshal(blob, &image); err != nil {
+		return nil, err
+	}
```
Comment (tianon) [inline daemon/containerd/image_history.go:46]: As a future enhancement, wouldn't it be better for this to use the content store's `ReaderAt` method and pass that directly to an `encoding/json`'s streaming interface so we don't have to have the full blob in memory _and_ the full parsed object? :sweat_smile:

File: daemon/containerd/image_history.go [hunk 7]
```diff
@@ -2,14 +2,104 @@ package containerd
...
+	for i := range diffIDs {
+		chainID := identity.ChainID(diffIDs[0 : i+1]).String()
+
+		use, err := s.Usage(ctx, chainID)
+		if err != nil {
+			return nil, err
+		}
+
+		sizes = append(sizes, use.Size)
+	}
```
Comment (tianon) [inline daemon/containerd/image_history.go:63]: "Size" of images in containerd is going to continue to be an interesting concept to grapple with, isn't it? :joy:

(Size of the content, for which platform, for which snapshotter, etc etc etc forever :sob:)

File: daemon/containerd/image_history.go [hunk 8]
```diff
@@ -2,14 +2,104 @@ package containerd
...
+
+	for _, h := range image.History {
+		size := int64(0)
+		if !h.EmptyLayer {
+			if len(sizes) == 0 {
+				return nil, errors.New("unable to find the size of the layer")
```
Comment (tianon) [inline daemon/containerd/image_history.go:69]: This is really a _mismatch_ between the number of layers we have vs the number of layers the image config claims we should, right?  (maybe the error could report that slightly better, including both numbers, or something?)

**Review by tianon — APPROVED** [review]
Mostly looks OK to me, just a few minor suggestions for where I think it could be improved (that IMO don't need to be in this PR).
